### PR TITLE
Added retry transformer to automatically retry network and S3 errors.

### PR DESCRIPTION
**Pull request by tkowalcz**
(cherry picked from commit e37ece1d6d556890cf573df19997c0c3af0751ec)

### CODE REVIEW

File: src/main/java/pl/codewise/amazon/client/AsyncS3Client.java [hunk 1]
```diff
@@ -167,7 +183,13 @@ private void listObjects(ListObjectsRequest listObjectsRequest, SingleEmitter<?
     }
 
     public Single<ObjectListing> listObjects(ListObjectsRequest listObjectsRequest) {
-        return Single.create(subscriber -> listObjects(listObjectsRequest, subscriber));
+        return Single.<ObjectListing>create(
+                subscriber -> listObjects(
+                        listObjectsRequest,
+                        subscriber)
```
Comment (tomasz-wiewiora) [inline src/main/java/pl/codewise/amazon/client/AsyncS3Client.java]: `)` in new line

File: src/main/java/pl/codewise/amazon/client/AsyncS3Client.java [hunk 2]
```diff
@@ -209,6 +231,19 @@ public void close() {
...
-        return Single.create(emitter -> retrieveResult(request, responseParser, emitter));
+        Single<T> single = Single.create(emitter ->
+                retrieveResult(
+                        request,
+                        responseParser,
+                        emitter
+                )
+        );
+
+        return single.compose(addRetries());
```
Comment (tomasz-wiewiora) [inline src/main/java/pl/codewise/amazon/client/AsyncS3Client.java]: inline single to be consistent

Comment (tomasz-wiewiora) [reply src/main/java/pl/codewise/amazon/client/AsyncS3Client.java]: also think about extracting single creation + compose with retries to some method with emitter as param.
Code like that is repeated above 3 times

File: src/main/java/pl/codewise/amazon/client/AsyncS3Client.java [hunk 3]
```diff
@@ -44,6 +47,11 @@
     public AsyncS3Client(ClientConfiguration configuration, NettyHttpClient httpClient) {
         this.httpClient = httpClient;
 
+        this.retryTransformer = createTransformerForRetryCount(
```
Comment (marcin-kapolka) [inline src/main/java/pl/codewise/amazon/client/AsyncS3Client.java]: I would not construct this within this ctor - use injection instead. It eases testing. Add a factory method which creates the transformer involving hidden constructor.

File: src/main/java/pl/codewise/amazon/client/AsyncS3Client.java [hunk 4]
```diff
@@ -209,6 +231,19 @@ public void close() {
...
+                        request,
+                        responseParser,
+                        emitter
+                )
+        );
+
+        return single.compose(addRetries());
+    }
+
+    @SuppressWarnings("unchecked")
```
Comment (marcin-kapolka) [inline src/main/java/pl/codewise/amazon/client/AsyncS3Client.java]: maybe move to another util class?

File: src/main/java/pl/codewise/amazon/client/AsyncS3Client.java (hunk 2 above)
Comment (tkowalcz) [reply src/main/java/pl/codewise/amazon/client/AsyncS3Client.java]: Will refactor.

File: src/main/java/pl/codewise/amazon/client/AsyncS3Client.java (hunk 3 above)
Comment (tkowalcz) [reply src/main/java/pl/codewise/amazon/client/AsyncS3Client.java]: Will refactor.

File: src/main/java/pl/codewise/amazon/client/AsyncS3Client.java (hunk 4 above)
Comment (tkowalcz) [reply src/main/java/pl/codewise/amazon/client/AsyncS3Client.java]: Will refactor.

**Pull request merged**